src/Agent.Worker/Release/Artifacts/TfGitArtifact.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker.Build;
+using Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts.Definition;
+using Microsoft.VisualStudio.Services.ReleaseManagement.WebApi.Contracts;
+using Newtonsoft.Json;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts
+{
+    public class TfGitArtifact : AgentService, IArtifactExtension
+    {
+        public Type ExtensionType => typeof(IArtifactExtension);
+        public AgentArtifactType ArtifactType => AgentArtifactType.TFGit;
+
+        public async Task DownloadAsync(IExecutionContext executionContext, ArtifactDefinition artifactDefinition, string downloadFolderPath)
+        {
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNull(artifactDefinition, nameof(artifactDefinition));
+            ArgUtil.NotNullOrEmpty(downloadFolderPath, nameof(downloadFolderPath));
+
+            var gitArtifactDetails = artifactDefinition.Details as GitArtifactDetails;
+            ArgUtil.NotNull(gitArtifactDetails, nameof(gitArtifactDetails));
+
+            ServiceEndpoint endpoint = executionContext.Endpoints.FirstOrDefault((e => string.Equals(e.Name, gitArtifactDetails.RepositoryId, StringComparison.OrdinalIgnoreCase)));
+            if (endpoint == null)
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMGitEndpointNotFound"));
+            }
+
+            var extensionManager = HostContext.GetService<IExtensionManager>();
+            ISourceProvider sourceProvider = (extensionManager.GetExtensions<ISourceProvider>()).FirstOrDefault(x => x.RepositoryType == WellKnownRepositoryTypes.TfsGit);
+
+            if (sourceProvider == null)
+            {
+                throw new InvalidOperationException(StringUtil.Loc("SourceArtifactNotFound", WellKnownRepositoryTypes.TfsGit));
+            }
+
+            executionContext.Variables.Set(Constants.Variables.Build.SourcesDirectory, downloadFolderPath);
```

Comment:
with this we might overwrite values of these variables set by service. can we pass these as parameters to the source provider? need to change this in GitHubArtifact impl. as well.